### PR TITLE
Fix git ls-files parsing across partial stream reads

### LINKED ISSUE #166: Fix git ls-files parsing across partial stream reads
## Summary

`GitLsFileOutputReader` searches for the NUL delimiter across the entire 4096-character buffer rather than only the characters returned by the current read.

## Impact

`git` standard output is pipe-backed and `TextReader.Read` may return fewer characters than requested before EOF. The unused or stale buffer tail can therefore be mistaken for a real NUL delimiter, splitting a single `git ls-files -z` record across multiple records. This can produce malformed paths or hashes, fail cache initialization, or leave files out of source-control-backed fingerprinting.

## Evidence

In `src/Common/SourceControl/GitFileHashProvider.cs`, `GitLsFileOutputReader.PopulateAsync` uses:

```csharp
Array.IndexOf(buffer, '\0', startIdx)
```

The search is not bounded by `readCnt`. Existing tests use `StringReader`, which does not exercise realistic partial pipe reads.

## Suggested fix

Bound delimiter searches to the valid range, for example with the `Array.IndexOf` overload that accepts a count. Add a test `TextReader` that deliberately returns small chunks and verify records spanning reads are reconstructed exactly.

## Acceptance criteria

- NUL-delimited records are parsed correctly regardless of read chunk boundaries.
- A test covers multiple partial reads, including a split within both staging metadata and a path.
- Existing source-control hashing tests continue to pass.

### PR DESCRIPTION
**Pull request by Evangelink**
## Summary

- bound NUL delimiter searches to the valid portion of each `TextReader` read
- reconstruct records exactly when staging metadata, paths, and delimiters span partial reads
- preserve the existing producer/consumer reader architecture and record-size policy
- add deterministic chunked-reader tests, including the 4095-character buffer edge

## Validation

- `GitFileHashProviderTests`: 11/11 passed on `net9.0`
- `GitFileHashProviderTests`: 11/11 passed on `net472`
- full Common test suite: 139/139 passed on `net9.0`

Closes #166

### CODE REVIEW

File: src/Common/SourceControl/GitFileHashProvider.cs [hunk 1]
```diff
@@ -315,72 +314,57 @@ internal List<string> ParseGitSubmoduleStatus(string output)
...
+
+        private readonly TextReader _reader;
+        private readonly char[] _buffer = new char[BufferSize];
+        private int _bufferPosition;
+        private int _bufferLength;
+        private bool _endOfStream;
 
         public GitLsFileOutputReader(TextReader reader)
         {
-            Task.Run(() => PopulateAsync(reader));
```
Comment (dfederm) [inline src/Common/SourceControl/GitFileHashProvider.cs:324]: Why the refactor? Did you perf test this change on a large repo? This code is perf tuned to support some of the largest repos at Microsoft. I'm hesitant to accept the refactor without seeing evidence that this is at least perf-neutral.

Additionally the refactor doesn't seem relevant to the bug fix, so my preference would be separating the fix from the refactor.

Comment (Evangelink) [reply src/Common/SourceControl/GitFileHashProvider.cs:324]: Good point. I removed the synchronous-reader refactor and restored the existing producer/consumer architecture; the source change is now limited to bounding the NUL scan to valid characters and fixing carry-over bookkeeping. I also changed the large test to cover the 4095-character buffer edge rather than expanding the supported record size, so this no longer changes the reader's execution or allocation model.


> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Review by dfederm — APPROVED** [review]

**Comment by Evangelink** [comment]
Thanks @dfederm
